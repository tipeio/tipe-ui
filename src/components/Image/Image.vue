<template>
  <div
    :data-tipe-ui="$options.name"
    :style="style"
  >
    <img
      :alt="alt"
      :src="url"
    >
  </div>
</template>

<script>
import vueTypes from 'vue-types'

export default {
  name: 'TipeImage',
  props: {
    alt: vueTypes.string.def(''),
    url: vueTypes.string.def('')
  },
  computed: {
    style() {
      return {
        backgroundImage: `url(${this.url})`
      }
    }
  }
}
</script>

<style scoped>
[data-tipe-ui='TipeImage'] {
  height: 100%;
  width: 100%;
  border: none;
  background-color: transparent;
  background: no-repeat center;
  background-size: cover;
}

/* hide the image visually, but not from screen readers */
img {
  position: absolute;
  width: 1px;
  height: 1px;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
  border: 0;
}

.waiting,
.error {
  display: flex;
  align-items: center;
  justify-content: center;
  height: 100%;
  width: 100%;
  background-color: rgba(0, 0, 0, 0.01);
  font-size: 0.8125rem;
  color: var(--text-gray);
}
</style>
