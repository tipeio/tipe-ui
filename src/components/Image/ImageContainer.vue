<template>
  <error-template v-if="status === 'error'"/>
  <tipe-image v-else-if="status === 'ok'"/>
  <waiting-template v-else/>
</template>

<script>
import vueTypes from 'vue-types'
import { Machine } from 'xstate'
import imageLoader from '../../libs/imageLoader'

import TipeImage from './Image.vue'
import ErrorTemplate from './templates/ErrorTemplate.vue'
import WaitingTemplate from './templates/WaitingTemplate.vue'

const status = Machine({
  initial: 'waiting',
  states: {
    waiting: {
      on: {
        done: 'ok',
        error: 'error'
      }
    },
    ok: {},
    error: {}
  }
})

export default {
  name: 'TipeImageContainer',
  components: {
    TipeImage,
    ErrorTemplate,
    WaitingTemplate
  },
  props: {
    alt: vueTypes.string.def(''),
    url: vueTypes.string.def(''),
    __imageLoader__: vueTypes.func.def(imageLoader)
  },
  data() {
    return {
      status: status.initial
    }
  },
  beforeMount() {
    this.load()
  },
  beforeUpdated() {
    this.load()
  },
  methods: {
    ok() {
      this.status = status.transition(this.status, 'done').value
    },
    error() {
      this.status = status.transition(this.status, 'error').value
    },
    load() {
      this.__imageLoader__(this.url)
        .then(this.ok)
        .catch(this.error)
    }
  }
}
</script>
