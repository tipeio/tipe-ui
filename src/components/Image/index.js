export { default } from './ImageContainer.vue'
