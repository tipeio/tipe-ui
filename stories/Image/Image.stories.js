import { storiesOf } from '@storybook/vue'
import { withKnobs, text } from '@storybook/addon-knobs/vue'
import { withMarkdownNotes } from '@storybook/addon-notes'

import TipeImage from '../../src/components/Image'

const notes = `
#### :props
:url | string |
:alt | string |
`

storiesOf('Image', module)
  .addDecorator(withKnobs)
  .add(
    'default',
    withMarkdownNotes(notes)(() => ({
      components: { TipeImage },
      data() {
        return {
          style: {
            height: '100px',
            width: '100px'
          },
          url: text('url', 'http://placekitten.com/200/200'),
          alt: text('alt', 'image of kitten')
        }
      },
      template: `
        <tipe-image
          :url="url"
          :alt="alt"
          :style="style"
        />`
    }))
  )
