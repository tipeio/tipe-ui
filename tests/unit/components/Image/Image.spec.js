import { shallowMount } from '@vue/test-utils'
import Image from '@/components/Image/Image.vue'
import faker from 'faker'

describe('Image', () => {
  describe('<template>', () => {
    it('matches previous snapshot', () => {
      const wrapper = shallowMount(Image)

      expect(wrapper).toMatchSnapshot()
    })

    it('has correct data-tipe-iu attibute', () => {
      const wrapper = shallowMount(Image)

      expect(wrapper.attributes()['data-tipe-ui']).toBe('TipeImage')
    })
  })

  describe(':props', () => {
    it(':url - image src attribute should match url', () => {
      const propsData = { url: faker.internet.url() }
      const wrapper = shallowMount(Image, { propsData })
      const img = wrapper.find('img')

      expect(img.attributes().src).toBe(propsData.url)
    })

    it(':alt - image alt attribute should match alt', () => {
      const propsData = { alt: faker.lorem.sentence() }
      const wrapper = shallowMount(Image, { propsData })
      const img = wrapper.find('img')

      expect(img.attributes().alt).toBe(propsData.alt)
    })
  })
})
