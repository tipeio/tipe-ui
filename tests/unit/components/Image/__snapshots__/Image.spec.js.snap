// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image <template> matches previous snapshot 1`] = `<div data-tipe-ui="TipeImage" style="background-image: url();"><img alt="" src=""></div>`;
